### PR TITLE
[Improve][Connector-V2][Jdbc-Source] Support for Decimal types as splict keys 

### PR DESCRIPTION
**Pull request by zhilinli123**
issues:https://github.com/apache/incubator-seatunnel/issues/4627

## Check list

* [ ] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/incubator-seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/incubator-seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-dist/pom.xml)
* [ ] Update the [`release-note`](https://github.com/apache/incubator-seatunnel/blob/dev/release-note.md).

### CODE REVIEW

**Comment by ic4y** [comment]
The current partition range is processed using Long, and bigint(20) will cause overflow, so it is necessary to modify the `PartitionParameter` related classes.

**Comment by ic4y** [comment]
Please update the document
```
partition_column [string]
The column name for parallelism's partition, only support numeric type.
```

**Comment by zhilinli123** [comment]
This is the document for now
image

**Comment by ic4y** [comment]
From the design perspective, it is necessary to support any DecimalType with scale=0. Otherwise, DecimalType(20,0), DecimalType(30,0), etc., will need to be modified. So it's better to do it well in one go

**Comment by ic4y** [comment]
> This is the document for now image

It is necessary to specify the supported data types clearly, such as Int, Long, etc

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-1/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc/JdbcMysqlIT.java [hunk 1]
```diff
@@ -60,7 +60,8 @@ public class JdbcMysqlIT extends AbstractJdbcIT {
             Lists.newArrayList(
                     "/jdbc_mysql_source_and_sink.conf",
                     "/jdbc_mysql_source_and_sink_parallel.conf",
-                    "/jdbc_mysql_source_and_sink_parallel_upper_lower.conf");
+                    "/jdbc_mysql_source_and_sink_parallel_upper_lower.conf",
+                    "/jdbc_mysql_source_and_sink_bu_parallel_upper_lower.conf");
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-1/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc/JdbcMysqlIT.java]: What is `bu`?

Comment (zhilinli123) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-1/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc/JdbcMysqlIT.java]: ` BIGINT unsigned` abbreviate

Comment (hailin0) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-1/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc/JdbcMysqlIT.java]: incomprehensible abbreviation

Comment (zhilinli123) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-1/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc/JdbcMysqlIT.java]: > sorry for my oversight

**Comment by zhilinli123** [comment]
> From the design perspective, it is necessary to support any DecimalType with scale=0. Otherwise, DecimalType(20,0), DecimalType(30,0), etc., will need to be modified. So it's better to do it well in one go

image
I have checked that Flink also uses Long type for storage. I think we do not need to change it for the time being
CC @hailin0 @ic4y

**Comment by ic4y** [comment]
> > From the design perspective, it is necessary to support any DecimalType with scale=0. Otherwise, DecimalType(20,0), DecimalType(30,0), etc., will need to be modified. So it's better to do it well in one go
> 
> image
> 
> I have checked that Flink also uses Long type for storage. I think we do not need to change it for the time being CC @hailin0 @ic4y

If Flink also supports bigint(20), it might have a similar issue.

**Review by ic4y — CHANGES_REQUESTED** [review]
Currently, there are two issues that need further discussion

1、The design needs to support any DecimalType with scale=0, instead of bigint(20).
2、Need to resolve the overflow issue of using Long to store partition boundaries

**Comment by zhilinli123** [comment]
> Currently, there are two issues that need further discussion
> 
> 1、The design needs to support any DecimalType with scale=0, instead of bigint(20). 2、Need to resolve the overflow issue of using Long to store partition boundaries

@TyrantLucifer @TaoZex @EricJoy2048 @hailin0 @Hisoka-X All bosses have time to help look at this thanks

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/source/JdbcSourceFactory.java [hunk 2]
```diff
@@ -200,7 +202,10 @@ static void validationPartitionColumn(String partitionColumn, SeaTunnelRowType r
     }
 
     private static boolean isNumericType(SeaTunnelDataType<?> type) {
-        return type.equals(BasicType.INT_TYPE) || type.equals(BasicType.LONG_TYPE);
+        // Currently, only numeric types are supported for reading partition keys
+        return type.equals(BasicType.INT_TYPE)
+                || type.equals(BasicType.LONG_TYPE)
+                || type.equals(new DecimalType(20, 0));
```
Comment (ic4y) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/source/JdbcSourceFactory.java]: The current implementation support any DecimalType with scale=0.

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-1/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc/JdbcMysqlIT.java [hunk 3]
```diff
@@ -64,6 +64,7 @@ public class JdbcMysqlIT extends AbstractJdbcIT {
     private static final String CREATE_SQL =
             "CREATE TABLE IF NOT EXISTS %s\n"
                     + "(\n"
+                    + "    `c_bigint_30`          BIGINT(40) unsigned   DEFAULT NULL,\n"
```
Comment (ic4y) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-1/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc/JdbcMysqlIT.java]: It is recommended to directly use DECIMAL(30,0) for partition field testing.

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-1/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc/JdbcMysqlIT.java [hunk 4]
```diff
@@ -194,11 +196,14 @@ Pair<String[], List<SeaTunnelRow>> initTestData() {
                 };
 
         List<SeaTunnelRow> rows = new ArrayList<>();
+        // Define a large bigint data value to test for compatibility above bigint(20)
+        BigDecimal testBigDecimalValue = new BigDecimal("2844674407371055000");
```
Comment (ic4y) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-1/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc/JdbcMysqlIT.java]: The number 2844674407371055000 is within the storage range of a long. You should use a value that is outside of long's storage capacity, such as 899,999,999,999,999,999,999,999,999,999.

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-1/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc/JdbcMysqlIT.java [hunk 5]
```diff
@@ -194,11 +196,14 @@ Pair<String[], List<SeaTunnelRow>> initTestData() {
...
         for (int i = 0; i < 100; i++) {
             byte byteArr = Integer.valueOf(i).byteValue();
             SeaTunnelRow row =
                     new SeaTunnelRow(
                             new Object[] {
+                                testBigDecimalValue.add(BigDecimal.valueOf(i)),
```
Comment (ic4y) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-1/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc/JdbcMysqlIT.java]: Do not delete the column c_bit_1.

**Review by ic4y — DISMISSED** [review]

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-1/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc/JdbcMysqlIT.java (hunk 5 above)
Comment (zhilinli123) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-1/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc/JdbcMysqlIT.java]: Done

Comment (zhilinli123) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-1/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc/JdbcMysqlIT.java]: Please have a look. Thank you. @ic4y

**Comment by zhilinli123** [comment]
@ic4y @hailin0  Please check it out. Thank you

**Comment by ic4y** [comment]
Wait for the CI to finish running

**Comment by zhilinli123** [comment]
CI success  @ic4y @EricJoy2048 @hailin0 Please check it out. Thank you

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-1/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc/JdbcMysqlIT.java [hunk 6]
```diff
@@ -64,47 +64,50 @@ public class JdbcMysqlIT extends AbstractJdbcIT {
...
-                    + "    `c_blob`               blob,\n"
-                    + "    `c_longblob`           longblob,\n"
-                    + "    `c_varbinary`          varbinary(255)        DEFAULT NULL,\n"
-                    + "    `c_binary`             binary(1)             DEFAULT NULL,\n"
-                    + "    `c_year`               year(4)               DEFAULT NULL,\n"
-                    + "    `c_int_unsigned`       int(10) unsigned      DEFAULT NULL,\n"
-                    + "    `c_integer_unsigned`   int(10) unsigned      DEFAULT NULL\n"
+                    + "    `c_bigint_30`            BIGINT(40)  unsigned  DEFAULT NULL,\n"
+                    + "    `c_decimal_unsigned_30`  DECIMAL(30) unsigned  DEFAULT NULL,\n"
+                    + "    `c_decimal_30`           DECIMAL(30)           DEFAULT NULL,\n"
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-1/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc/JdbcMysqlIT.java]: move to end

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-1/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc/JdbcMysqlIT.java [hunk 7]
```diff
@@ -150,6 +153,9 @@ String driverUrl() {
     Pair<String[], List<SeaTunnelRow>> initTestData() {
         String[] fieldNames =
                 new String[] {
+                    "c_bigint_30",
+                    "c_decimal_unsigned_30",
+                    "c_decimal_30",
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-1/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc/JdbcMysqlIT.java]: move to end

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-1/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc/JdbcMysqlIT.java [hunk 8]
```diff
@@ -194,11 +200,16 @@ Pair<String[], List<SeaTunnelRow>> initTestData() {
...
+        BigDecimal bigintValue = new BigDecimal("2844674407371055000");
+        BigDecimal decimalValue = new BigDecimal("999999999999999999999999999899");
         for (int i = 0; i < 100; i++) {
             byte byteArr = Integer.valueOf(i).byteValue();
             SeaTunnelRow row =
                     new SeaTunnelRow(
                             new Object[] {
+                                bigintValue.add(BigDecimal.valueOf(i)),
+                                decimalValue.add(BigDecimal.valueOf(i)),
+                                decimalValue.add(BigDecimal.valueOf(i)),
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-1/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc/JdbcMysqlIT.java]: move to end

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-1/src/test/resources/jdbc_mysql_source_and_sink.conf [hunk 9]
```diff
@@ -42,12 +42,12 @@ sink {
         user = "root"
         password = "Abc!@#135_seatunnel"
 
-        query = """insert into sink (c_bit_1, c_bit_8, c_bit_16, c_bit_32, c_bit_64, c_boolean, c_tinyint, c_tinyint_unsigned, c_smallint, c_smallint_unsigned,
+        query = """insert into sink (c_bigint_30,c_decimal_unsigned_30,c_decimal_30,c_bit_1, c_bit_8, c_bit_16, c_bit_32, c_bit_64, c_boolean, c_tinyint, c_tinyint_unsigned, c_smallint, c_smallint_unsigned,
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-1/src/test/resources/jdbc_mysql_source_and_sink.conf]: move to end

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-1/src/test/resources/jdbc_mysql_source_and_sink_parallel_upper_lower.conf [hunk 10]
```diff
@@ -46,12 +46,12 @@ sink {
         user = "root"
         password = "Abc!@#135_seatunnel"
         connection_check_timeout_sec = 100
-        query = """insert into sink (c_bit_1, c_bit_8, c_bit_16, c_bit_32, c_bit_64, c_boolean, c_tinyint, c_tinyint_unsigned, c_smallint, c_smallint_unsigned,
+        query = """insert into sink (c_bigint_30,c_decimal_unsigned_30,c_decimal_30,c_bit_1, c_bit_8, c_bit_16, c_bit_32, c_bit_64, c_boolean, c_tinyint, c_tinyint_unsigned, c_smallint, c_smallint_unsigned,
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-1/src/test/resources/jdbc_mysql_source_and_sink_parallel_upper_lower.conf]: move to end

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-1/src/test/resources/jdbc_mysql_source_and_sink_parallel.conf [hunk 11]
```diff
@@ -45,12 +45,12 @@ sink {
         user = "root"
         password = "Abc!@#135_seatunnel"
         connection_check_timeout_sec = 100
-        query = """insert into sink (c_bit_1, c_bit_8, c_bit_16, c_bit_32, c_bit_64, c_boolean, c_tinyint, c_tinyint_unsigned, c_smallint, c_smallint_unsigned,
+        query = """insert into sink (c_bigint_30,c_decimal_unsigned_30,c_decimal_30,c_bit_1, c_bit_8, c_bit_16, c_bit_32, c_bit_64, c_boolean, c_tinyint, c_tinyint_unsigned, c_smallint, c_smallint_unsigned,
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-1/src/test/resources/jdbc_mysql_source_and_sink_parallel.conf]: move to end

Comment (zhilinli123) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-1/src/test/resources/jdbc_mysql_source_and_sink_parallel.conf]: Thanks for checking. I fixed it
@hailin0

**Comment by zhilinli123** [comment]
PTAL:@ic4y @Hisoka-X @hailin0  thks

**Review by ic4y — DISMISSED** [review]
LGTM

**Comment by zhilinli123** [comment]
PTAL:@EricJoy2048 @hailin0  thks
I have solved some code conflicts, please help to open CI, thank you  
image

**Comment by zhilinli123** [comment]
I have solved some code conflicts, please help to open CI, thank you

**Comment by zhilinli123** [comment]
I have solved some code conflicts, please help to open CI, thank you

**Comment by hailin0** [comment]
update title

**Comment by zhilinli123** [comment]
> update title
in this way?

**Review by ic4y — APPROVED** [review]

**Review by davidzollo — APPROVED** [review]
+1

**Comment by zhilinli123** [comment]
For the current pr- I will modify all source-documents related to boundary value types - in jdbc

**Pull request merged**